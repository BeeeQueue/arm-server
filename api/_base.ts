import type { JsonValue } from "type-fest"

import { Boom, internal, isBoom } from "@hapi/boom"
import * as Sentry from "@sentry/node"
import type { VercelRequest, VercelResponse } from "@vercel/node"

import { Logger } from "./_logger"

import "@sentry/tracing"

type Handler = (request: VercelRequest, response: VercelResponse) => Promise<void> | void
type CustomHandlerResponse<Body extends JsonValue = JsonValue> = Body | Boom<null> | undefined
export type CustomHandler<Body extends JsonValue = JsonValue> = (
  request: VercelRequest,
  response: Pick<VercelResponse, "setHeader">,
) => Promise<CustomHandlerResponse<Body>> | CustomHandlerResponse<Body>

Sentry.init({
  enabled: process.env.VERCEL_ENV !== "development" && !!process.env.SENTRY_DSN,
  release: process.env.VERCEL_GIT_COMMIT_SHA,
  dsn: process.env.SENTRY_DSN,
  environment: process.env.VERCEL_ENV as string,
  integrations: [new Sentry.Integrations.Http({ tracing: true })],
  tracesSampleRate: 0.5,
})

Sentry.setTag("app", "api")

export const createHandler =
  (name: string, handler: CustomHandler): Handler =>
  async (req, res) => {
    let result: CustomHandlerResponse

    const trx = Sentry.startTransaction({
      name,
      op: "transaction",
    })

    try {
      result = (await handler(req, res)) ?? null
    } catch (error) /* istanbul ignore next */ {
      result = internal(error.message)

      Sentry.setContext("response", {
        status: res.statusCode,
      })

      Logger.error(error)
      Sentry.captureException(error)
    }

    trx.finish()

    await Sentry.flush(1000)

    res.setHeader("Content-Type", "application/json")

    if (isBoom(result)) {
      const { payload, statusCode, headers } = result.output
      const body = { ...payload, ok: false }

      /* istanbul ignore next */
      for (const [key, value] of Object.entries(headers)) {
        res.setHeader(key, value!)
      }

      res.status(statusCode).json(body)
      Logger.debug(
        `${req.method as string} ${name} ${res.statusCode}\n${JSON.stringify(body, null, 2)}`,
      )

      return
    }

    const body = result ?? "null"

    res.json(body)
    Logger.debug(
      `${req.method as string} ${name} ${res.statusCode}\n${JSON.stringify(body, null, 2)}`,
    )
  }
