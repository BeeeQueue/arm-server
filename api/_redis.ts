import { URL } from "url"

import IORedis, { RedisOptions } from "ioredis"

const redisUrl = new URL(process.env.REDIS_URL as string)
const redisConfig: RedisOptions = {
  host: redisUrl.hostname,
  port: Number(redisUrl.port),
  username: redisUrl.username,
  password: redisUrl.password,
  tls: redisUrl.protocol === "rediss:" ? {} : undefined,
}

export const Redis = new IORedis(redisConfig)
