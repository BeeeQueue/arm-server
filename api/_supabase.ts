import { SupabaseClient } from "@supabase/supabase-js"

export const Supabase = new SupabaseClient(
  process.env.SUPABASE_URL as string,
  process.env.SUPABASE_SERVICE_TOKEN as string,
)
