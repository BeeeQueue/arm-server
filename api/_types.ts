export type OfflineDatabaseEntry = {
  sources: string[]
  type: string
  title: string
  picture: string
  relations: string[]
  thumbnail: string
  episodes: number
  synonyms: string[]
}

export type OfflineDatabaseData = {
  data: OfflineDatabaseEntry[]
}

export enum Source {
  AniList = "anilist",
  AniDB = "anidb",
  MAL = "myanimelist",
  Kitsu = "kitsu",
}

export type Relation = {
  [key in Source]: number | null
}

export type QueryParamInput = {
  source: Source
  id: number
}

export type BodyItem = {
  [key in Source]?: number
}

export type BodyInput = BodyItem | BodyItem[]
