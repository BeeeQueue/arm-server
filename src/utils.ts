/* eslint-disable @typescript-eslint/no-explicit-any */

export const enumToArray = <E>(e: E): E[] => Object.values(e)

export const isEmpty = <T extends Record<string, unknown> | Record<string, unknown>[]>(
  obj?: T | null | undefined,
): obj is T => {
  if (obj == null) return true

  if (Array.isArray(obj)) return obj.length === 0

  return Object.keys(obj).length === 0
}
