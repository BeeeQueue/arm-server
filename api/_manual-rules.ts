import { captureMessage } from "@sentry/node"

import { Relation } from "./_types"

const rules = {
  // Kaguya-sama
  "anidb:14111": "anilist:101921",
  // Shield Hero
  "anidb:13246": "anilist:99263",
  // Vivy #385
  "anidb:15988": "myanimelist:46095",
  // Osananajimi #386
  "anidb:15756": "myanimelist:43007",
  // Iruma-kun S2 #387
  "anidb:15428": "myanimelist:41402",
}

const parseRule = (rule: string) => {
  const [source, id] = rule.split(":") as [keyof Relation, string]

  return [source, Number(id)] as const
}

export const updateBasedOnManualRules = (relations: Relation[]) => {
  const newRelations = [...relations]

  Object.entries(rules).forEach(([from, to]) => {
    const [fromSource, fromId] = parseRule(from)
    const [toSource, toId] = parseRule(to)

    const fromIndex = relations.findIndex((relation) => relation[fromSource] === fromId)
    if (fromIndex < 0) {
      captureMessage(`Could not find rule source for ${from} -> ${to}!`)
    }

    const toIndex = relations.findIndex((relation) => relation[toSource] === toId)

    // Remove the old one from the new array
    newRelations.splice(toIndex, 1)
    // Update the correct entry with the new data
    newRelations[fromIndex] = Object.assign(newRelations[fromIndex], relations[toIndex])

    console.log(relations[fromIndex], newRelations[fromIndex])
  })

  return newRelations
}
