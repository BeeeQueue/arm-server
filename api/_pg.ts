import KnexBuilder, { Knex as IKnex } from "knex"

import { config } from "../knexfile"

import { Relation } from "./_types"

declare module "knex/types/tables" {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Tables {
    relations: Relation
    // eslint-disable-next-line @typescript-eslint/naming-convention
    relations_composite: IKnex.CompositeTableType<Relation, Relation, Relation>
  }
}

const env = (process.env.NODE_ENV as keyof typeof config) || "production"
export const Knex = KnexBuilder(config[env])
