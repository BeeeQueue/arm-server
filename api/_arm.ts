import fetch from "node-fetch"

import { captureException } from "@sentry/node"

import { Logger } from "./_logger"
import { updateBasedOnManualRules } from "./_manual-rules"
import { Knex } from "./_pg"
import {
  BodyItem,
  OfflineDatabaseData,
  OfflineDatabaseEntry,
  QueryParamInput,
  Relation,
  Source,
} from "./_types"

const isDefined = <T>(obj: T | null | undefined): obj is T => obj != null

const chunk = <T>(arr: T[], size: number) =>
  arr.reduce((all, one, i) => {
    const ch = Math.floor(i / size)
    all[ch] = [...(all[ch] ?? []), one]
    return all
  }, [] as T[][])

export class ArmData {
  private static regexes = {
    anilist: /anilist.co\/anime\/(\d+)$/,
    anidb: /anidb.net\/a(?:nime\/)?(\d+)$/,
    mal: /myanimelist.net\/anime\/(\d+)$/,
    kitsu: /kitsu.io\/anime\/(.+)$/,
  }

  private static formatEntry = (entry: OfflineDatabaseEntry): Relation | null => {
    const relation: Relation = {
      anilist: null,
      anidb: null,
      kitsu: null,
      myanimelist: null,
    }

    entry.sources.forEach((src) => {
      const anilistMatch = ArmData.regexes.anilist.exec(src)
      if (anilistMatch) {
        const id = Number(anilistMatch[1])

        if (isNaN(id)) throw new Error(`${src}'s ID is not a number!!`)

        relation.anilist = id
      }

      const anidbMatch = ArmData.regexes.anidb.exec(src)
      if (anidbMatch) {
        const id = Number(anidbMatch[1])

        if (isNaN(id)) throw new Error(`${src}'s ID is not a number!!`)

        relation.anidb = id
      }

      const malMatch = ArmData.regexes.mal.exec(src)
      if (malMatch) {
        const id = Number(malMatch[1])

        if (isNaN(id)) throw new Error(`${src}'s ID is not a number!!`)

        relation.myanimelist = id
      }

      const kitsuMatch = ArmData.regexes.kitsu.exec(src)
      if (kitsuMatch) {
        const id = Number(kitsuMatch[1])

        if (isNaN(id)) throw new Error(`${src}'s ID is not a number!!`)

        relation.kitsu = id
      }
    })

    if (Object.values(relation).every((v) => !v)) {
      return null
    }

    return relation
  }

  private static pushToSupabase = async (relations: Relation[]) => {
    const chunks = chunk(relations, 5_000)
    Logger.debug(`Pushing ${relations.length} relations, split into ${chunks.length} chunks...`)

    const promises = chunks.map((chunkedRelations) =>
      Knex.from("relations")
        .insert(chunkedRelations)
        .catch((err) => err),
    )

    const results = await Promise.allSettled(promises)
    const rejections = results.filter(
      (result): result is PromiseRejectedResult => result.status === "rejected",
    )

    if (rejections.length > 0) {
      Logger.error(rejections.map((rejection) => rejection.reason))
    }
  }

  private static fetchDatabase = async (): Promise<OfflineDatabaseEntry[] | null> => {
    const response = await fetch(
      "https://raw.githubusercontent.com/manami-project/anime-offline-database/master/anime-offline-database.json",
    ).catch((err: Error) => err)

    if (response instanceof Error || response.status >= 400) {
      const error = new Error(
        `Failed to fetch updated database!\n${
          response instanceof Error ? response.message : response.status
        }`,
      )

      Logger.error(error)
      captureException(error)

      return null
    }

    return ((await response.json()) as OfflineDatabaseData).data
  }

  public static updateDatabase = async () => {
    Logger.info("Fetching updated Database...")
    const data = await ArmData.fetchDatabase()
    Logger.info("Fetched updated Database.")

    if (data == null) {
      Logger.info("got no data")
      return
    }

    Logger.info("Formatting data...")
    const entries = data.map(ArmData.formatEntry).filter(isDefined)
    Logger.info("Formatted data.")

    Logger.info("Executing manual rules...")
    const fixedEntries = updateBasedOnManualRules(entries)

    Logger.info("Uploading to Supabase...")
    await ArmData.pushToSupabase(fixedEntries)

    Logger.info("Done.")
  }

  public static getRelation = async (input: QueryParamInput) => {
    const result: Relation[] | Error = await Knex.from("relations")
      .select(Object.values(Source))
      .where(input.source, input.id)
      .limit(1)
      .catch((err: Error) => err)

    if (result instanceof Error) {
      Logger.error(result)

      return null
    }

    return result?.[0] ?? null
  }

  public static getRelations = async (input: BodyItem[]) => {
    const result: Relation[] | Error = await Knex("relations")
      .where(function () {
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        input.forEach((item) => this.orWhere(item))
      })
      .select(Object.values(Source))
      .catch((err: Error) => err)

    if (result instanceof Error) {
      Logger.error(result)

      return null
    }

    if (result == null) return []

    return input.map((item) => {
      const realItem = Object.entries(item)[0] as [Source, number]

      return result.find((relation) => relation[realItem[0]] === realItem[1]) ?? null
    })
  }
}
