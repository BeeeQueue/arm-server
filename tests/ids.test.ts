import type { Server } from "http"
import { URLSearchParams } from "url"

import HttpClient, { Got, Response } from "got"
import listen from "test-listen"
import type { JsonValue } from "type-fest"
import { createServer } from "vercel-node-server"

import { Relation, Source } from "../api/_types"
import IdFunction from "../api/ids"

import { cleanUpDb, insertRelations } from "./utils"

const createSnapshot = (
  input: { json?: unknown; searchParams?: URLSearchParams },
  { url, method, statusCode, headers: { date, ...headers }, body }: Response,
) => ({
  request: {
    url: url.replace(/localhost:\d+/, "localhost"),
    ...input,
  },
  response: {
    method,
    statusCode,
    headers,
    body,
  },
})

let server: Server
let baseUrl: string
let client: Got

beforeAll(async () => {
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  server = createServer(IdFunction)
  baseUrl = await listen(server)

  client = HttpClient.extend({
    prefixUrl: baseUrl,
    responseType: "json",
    allowGetBody: true,
    throwHttpErrors: false,
  })
})

beforeEach(async () => {
  await cleanUpDb()

  await insertRelations([{ anilist: 1337 }])
})

afterAll(
  () =>
    new Promise((resolve) => {
      server.close(resolve)
    }),
)

describe("query params", () => {
  test("returns 200 with null if not found", async () => {
    const searchParams = new URLSearchParams({
      source: Source.AniList,
      id: (1338).toString(),
    })

    const response = await client.get("api/ids", {
      searchParams,
      headers: {
        accept: "application/json",
      },
    })

    expect(createSnapshot({ searchParams }, response)).toMatchSnapshot()

    expect(response.headers["content-type"]).toBe("application/json")
    expect(response.body).toBe(null)
  })

  test("returns 200 with relation if found", async () => {
    const searchParams = new URLSearchParams({
      source: Source.AniList,
      id: (1337).toString(),
    })

    const response = await client.get<Relation | null>("api/ids", {
      searchParams,
    })

    expect(createSnapshot({ searchParams }, response)).toMatchSnapshot()
  })
})

describe("json body", () => {
  const methods = ["get", "post"] as const

  describe("single", () => {
    methods.forEach((method) => {
      describe(method.toUpperCase(), () => {
        test("returns 200 with null if not found", async () => {
          const json = {
            [Source.AniList]: 1338,
          }

          const response = await client[method]("api/ids", {
            json,
            headers: {
              accept: "application/json",
            },
          })

          expect(createSnapshot({ json }, response)).toMatchSnapshot()
        })

        test("returns 200 with relation if found", async () => {
          const json = {
            [Source.AniList]: 1337,
          }

          const response = await client[method]("api/ids", {
            json,
            headers: {
              accept: "application/json",
            },
          })

          expect(createSnapshot({ json }, response)).toMatchSnapshot()
        })
      })
    })
  })

  describe("multiple", () => {
    beforeEach(async () => {
      await insertRelations([{ anidb: 42 }])
    })

    methods.forEach((method) => {
      describe(method.toUpperCase(), () => {
        test("returns 200 with null if not found", async () => {
          const json = [
            {
              [Source.AniList]: 1338,
            },
            {
              [Source.AniDB]: 42,
            },
          ]

          const response = await client[method]("api/ids", {
            json,
            headers: {
              accept: "application/json",
            },
          })

          expect(createSnapshot({ json }, response)).toMatchSnapshot()
        })

        test("returns 200 with relation if found", async () => {
          const json = [
            {
              [Source.AniList]: 1337,
            },
            {
              [Source.AniDB]: 42,
            },
          ]

          const response = await client[method]("api/ids", {
            json,
            headers: {
              accept: "application/json",
            },
          })

          expect(createSnapshot({ json }, response)).toMatchSnapshot()
        })
      })
    })
  })
})

describe("input errors", () => {
  const inputs: Array<["searchParams" | "json", JsonValue]> = [
    ["searchParams", { source: "aniList", id: 1337 }],
    ["searchParams", { source: "anilist", id: -1 }],
    ["searchParams", { source: "anilist" }],
    ["searchParams", { id: 1337 }],
    ["json", {}],
    ["json", { aniList: 1337 }],
    ["json", { anilist: -1 }],
    ["json", { anilist: 1.5 }],
    ["json", []],
    ["json", [{}]],
    ["json", [{ aniList: 1337 }]],
    ["json", [{ anilist: -1 }]],
  ]

  test.each(inputs)("%s %p", async (inputType, input) => {
    const formattedInput = inputType === "searchParams" ? new URLSearchParams(input) : input

    const response = await client("api/ids", {
      [inputType]: formattedInput,
      headers: {
        accept: "application/json",
      },
    })

    expect(createSnapshot({ [inputType]: input }, response)).toMatchSnapshot()
    expect(response.statusCode).toBe(400)
  })
})
