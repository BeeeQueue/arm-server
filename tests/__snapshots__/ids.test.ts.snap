// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json body single (get) returns 200 with null if not found 1`] = `
Object {
  "request": Object {
    "body": Object {
      "anilist": 1338,
    },
    "url": "http://localhost/api/ids",
  },
  "response": Object {
    "body": null,
    "headers": Object {
      "connection": "close",
      "content-length": "4",
      "content-type": "application/json",
    },
    "method": null,
    "statusCode": 200,
  },
}
`;

exports[`json body single (get) returns 200 with relation if found 1`] = `
Object {
  "request": Object {
    "body": Object {
      "anilist": 1337,
    },
    "url": "http://localhost/api/ids",
  },
  "response": Object {
    "body": Object {
      "anidb": null,
      "anilist": 1337,
      "kitsu": null,
      "myanimelist": null,
    },
    "headers": Object {
      "connection": "close",
      "content-length": "61",
      "content-type": "application/json",
    },
    "method": null,
    "statusCode": 200,
  },
}
`;

exports[`json body single (post) returns 200 with null if not found 1`] = `
Object {
  "request": Object {
    "body": Object {
      "anilist": 1338,
    },
    "url": "http://localhost/api/ids",
  },
  "response": Object {
    "body": null,
    "headers": Object {
      "connection": "close",
      "content-length": "4",
      "content-type": "application/json",
    },
    "method": null,
    "statusCode": 200,
  },
}
`;

exports[`json body single (post) returns 200 with relation if found 1`] = `
Object {
  "request": Object {
    "body": Object {
      "anilist": 1337,
    },
    "url": "http://localhost/api/ids",
  },
  "response": Object {
    "body": Object {
      "anidb": null,
      "anilist": 1337,
      "kitsu": null,
      "myanimelist": null,
    },
    "headers": Object {
      "connection": "close",
      "content-length": "61",
      "content-type": "application/json",
    },
    "method": null,
    "statusCode": 200,
  },
}
`;

exports[`query params returns 200 with null if not found 1`] = `
Object {
  "request": Object {
    "searchParams": URLSearchParams {
      Symbol(query): Array [
        "source",
        "anilist",
        "id",
        "1338",
      ],
      Symbol(context): null,
    },
    "url": "http://localhost/api/ids?source=anilist&id=1338",
  },
  "response": Object {
    "body": null,
    "headers": Object {
      "connection": "close",
      "content-length": "4",
      "content-type": "application/json",
    },
    "method": null,
    "statusCode": 200,
  },
}
`;

exports[`query params returns 200 with relation if found 1`] = `
Object {
  "request": Object {
    "searchParams": URLSearchParams {
      Symbol(query): Array [
        "source",
        "anilist",
        "id",
        "1337",
      ],
      Symbol(context): null,
    },
    "url": "http://localhost/api/ids?source=anilist&id=1337",
  },
  "response": Object {
    "body": Object {
      "anidb": null,
      "anilist": 1337,
      "kitsu": null,
      "myanimelist": null,
    },
    "headers": Object {
      "connection": "close",
      "content-length": "61",
      "content-type": "application/json",
    },
    "method": null,
    "statusCode": 200,
  },
}
`;
